server/src/main/java/com/octopus/teamcity/opentelemetry/server/OTELHelper.java

```diff
@@ -0,0 +1,101 @@
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+
+public class OTELHelper {
+
+    private final OpenTelemetry openTelemetry;
+    private final Tracer tracer;
+    private final HashMap<String, Span> spanMap;
+
+    public OTELHelper(Map<String, String> headers, String exporterEndpoint) {
+
+        SpanProcessor spanProcessor = buildSpanProcessor(headers, exporterEndpoint);
+
+        Resource serviceNameResource = Resource
+                .create(Attributes.of(ResourceAttributes.SERVICE_NAME, PluginConstants.SERVICE_NAME));
+        SdkTracerProvider sdkTracerProvider = SdkTracerProvider.builder()
+                .setResource(Resource.getDefault().merge(serviceNameResource))
+                .addSpanProcessor(spanProcessor)
+                .build();
+        this.openTelemetry = OpenTelemetrySdk.builder()
+                .setTracerProvider(sdkTracerProvider)
+                .setPropagators(ContextPropagators.create(W3CTraceContextPropagator.getInstance()))
+                .buildAndRegisterGlobal();
+        Loggers.SERVER.info("OTEL_PLUGIN: OTEL_PLUGIN: OpenTelemetry plugin started.");
+        this.tracer = this.openTelemetry.getTracer(PluginConstants.TRACER_INSTRUMENTATION_NAME);
+        this.spanMap = new HashMap<>();
+    }
+
+    private SpanProcessor buildSpanProcessor(Map<String, String> headers, String exporterEndpoint) {
+        OtlpGrpcSpanExporterBuilder spanExporterBuilder = OtlpGrpcSpanExporter.builder();
+        headers.forEach(spanExporterBuilder::addHeader);
+        spanExporterBuilder.setEndpoint(exporterEndpoint);
+        SpanExporter spanExporter = spanExporterBuilder.build();
+
+        Loggers.SERVER.debug("OTEL_PLUGIN: Opentelemetry export headers: " + LogMasker.mask(headers.toString()));
+        Loggers.SERVER.debug("OTEL_PLUGIN: Opentelemetry export endpoint: " + exporterEndpoint);
+
+        return BatchSpanProcessor.builder(spanExporter).build();
+    }
+
+    public boolean pluginReady() {
+        return this.openTelemetry != null;
+    }
+
+    public Span getParentSpan(String buildId) {
+        return this.spanMap.computeIfAbsent(buildId, key -> this.tracer.spanBuilder(buildId).startSpan());
+    }
+
+    public Span createSpan(String buildId, Span parentSpan) {
+        return this.spanMap.containsKey(buildId) ?
+                this.spanMap.get(buildId) :
+                this.tracer.spanBuilder(buildId).setParent(Context.current().with(parentSpan)).startSpan();
+    }
+
+    public Span createChildSpan(Span parentSpan, String spanName, long startTime) {
+        Loggers.SERVER.info("OTEL_PLUGIN: Creating child span " + spanName + " under parent " + parentSpan);
+        return this.tracer.spanBuilder(spanName)
+                .setParent(Context.current().with(parentSpan))
+                .setStartTimestamp(startTime, TimeUnit.MILLISECONDS)
+                .startSpan();
+    }
+
+    public void addSpanToMap(String buildId, Span span) {
```

Comment:
I'd suggest removing the `ToMap`/`FromMap` part of these names - it's not relevant to the caller to know that it's backed by a HashMap.

Reply:
nws can be removed.